[GPU Process] [Filters] Cache the SVGFilter applying results in RemoteResourceCache

shallawa opened the pull request:
#### b686ba471d7aed8eb470f10f048653da3c5e4b80
<pre>
[GPU Process] [Filters] Cache the SVGFilter applying results in RemoteResourceCache
<a href="https://bugs.webkit.org/show_bug.cgi?id=232845">https://bugs.webkit.org/show_bug.cgi?id=232845</a>
rdar://85426641

Reviewed by Simon Fraser.

This allows caching the results of applying an SVGFilter to a source ImageBuffer
in GPU Process. We should be able to use the cached result if the Filter was not
changed. We should be also able to clear some of the result FilterImages if a
FilterEffect was changed.

* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::mergeEffects):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBufferInternal):
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawFilteredImageBuffer):

Canonical link: <a href="https://commits.webkit.org/264087@main">https://commits.webkit.org/264087@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48a0df23c010b89a37812df49d23b1401a35ed77

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6677 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6893 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8266 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/6950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6839 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7326 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8355 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6038 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6119 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8820 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6608 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5414 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6003 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1572 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10175 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/6376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilterTransaction.h:
```diff
@@ -0,0 +1,41 @@
+class FilterEffect;
+
+struct SVGFilterTransactionItem {
+    unsigned index;
```
Index into what? This isn't the index in this Vector<>, right?

It's a bit confusing to use these vectors for SVG filters, where the filter structure is a graph, and more confusing to have this standalone structure that contains indexes into something else.

shallawa replied:
It is an index to the `SVGFilter::m_effects` which is a Vector of the unique effects in the `SVGFilter` graph. The order of the effects in WebProcess and GPUProcess is the same. This is why when a certain `FilterEffect` changes in WebProcess we lookup its index in `SVGFilter::m_effects` and send it with this index to GPUProcess in `SVGFilterTransaction`. Using this index, GPUProcess will replace this effect in `SVGFilter::m_effects` and delete its `FilterImage` along with the result of all the dependent `FilterEffects`.

shallawa replied:
`SVGFilterExpression` represents the graph of the `SVGFilter`. It is a Vector of `SVGFilterExpressionTerm`. Each `SVGFilterExpressionTerm` has an index pointing to the `SVGFilter::m_effects`.

smfr replied:
Is there a way we can package up SVGFilterExpression and SVGFilterTransaction into one struct, so we don't have these dangling indexes?

shallawa replied:
Yes we can send the whole SVGFilter to GPUP and merge the new effects with the old cached one. If an effect does not match its cached corresponding one, we replace the old with the new and we clear the result and all the dependent results of this effect.

smfr commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -253,8 +254,39 @@ void RemoteDisplayListRecorder::drawFilteredImageBuffer(std::optional<RenderingR
+        return;
+    }
+
+    RefPtr cahcedFilter = resourceCache().cachedFilter({ filter->renderingResourceIdentifier(), m_webProcessIdentifier });
```
"cahcedFilter"

smfr commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -253,8 +254,39 @@ void RemoteDisplayListRecorder::drawFilteredImageBuffer(std::optional<RenderingR
+    }
+
+    RefPtr cahcedFilter = resourceCache().cachedFilter({ filter->renderingResourceIdentifier(), m_webProcessIdentifier });
+    if (!cahcedFilter || !is<SVGFilter>(*cahcedFilter)) {
```
This could be a single `dynamicDowncast<>`

shallawa committed:
[GPU Process] [Filters] Cache the SVGFilter applying results in RemoteResourceCache
https://bugs.webkit.org/show_bug.cgi?id=232845
rdar://85426641

Reviewed by Simon Fraser.

This allows caching the results of applying an SVGFilter to a source ImageBuffer
in GPU Process. We should be able to use the cached result if the Filter was not
changed. We should be also able to clear some of the result FilterImages if a
FilterEffect was changed.

* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::mergeEffects):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBufferInternal):
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawFilteredImageBuffer):

Canonical link: https://commits.webkit.org/264087@main
